sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -154,8 +154,12 @@ class SimpleTestOptimizer extends Optimizer(
 object PushProjectThroughSample extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
     // Push down projection into sample
-    case Project(projectList, Sample(lb, up, replace, seed, child)) =>
-      Sample(lb, up, replace, seed, Project(projectList, child))()
+    case proj @ Project(projectList, Sample(lb, up, replace, seed, child)) =>
+      if (!projectList.exists(_.find(!_.deterministic).nonEmpty)) {
```

Comment:
I think, even when `replace=true`, we can push projects down if all expressions in projects are deterministic.
But, when `replace=false`, we can always push down, so istm we need to add `replace` in the condition like;
```
if (!replace || !projectList.exists(_.find(!_.deterministic).nonEmpty)) {
```